### PR TITLE
Avoid extra parens for wildcard type signature

### LINKED ISSUE #2674: Use type signature: Don't insert excess parentheses .
When completing a type signature HLS is very conservative and always adds parentheses around the actual type.

For example something like `myHead :: [a] -> _` will get a type of `myHead :: [a] -> (a)`. Note the parentheses around the a.

While I personally don't have a big problem with this. Many of my colleagues insist on them being removed. Which usually makes these signatures easier to read!

**Describe the solution you'd like**

Ideally I would like the inserted type signature to be the one with the least parentheses. For example `Just _` will require braces if the inserted type is `Either Int Char`. But if it's just `Char` there is no need for them. But in something like `a -> _ -> ...` then we can get away without braces for `Either Int Char`!

It's likely not trivial to achieve this. But it would be really nice.

### PR DESCRIPTION
**Pull request by xsebek**
Avoid parentheses in simple cases (`a`, `Char`, `()`,...).

The good news is, that getting types inside tuples and lists will not force you to apply hint "Redundant brackets":
```haskell
g :: a -> _
g x = (x, const x)
--------------------------
-- before
g :: a -> ((a, b -> a))
--------------------------
-- after
g :: a -> (a, b -> a)
```

The caveat is that it is still context unaware (works on text only) and the common use case of `f :: _` is not improved.

- Hopefully closes #2674 

<a href="https://gitpod.io/#https://github.com/haskell/haskell-language-server/pull/2764"></a>

### CODE REVIEW

**Comment by michaelpj** [comment]
Could you add some tests for the examples you give in the PR description?

**Review by michaelpj — APPROVED** [review]
LGTM modulo adding some tests.

**Comment by xsebek** [comment]
@michaelpj there are now tests for the tuple, list and unit type. :slightly_smiling_face:

**Comment by xsebek** [comment]
(circleci did not finish so I rebased on master to give it a second chance)

File: ghcide/test/exe/Main.hs [hunk 1]
```diff
@@ -1188,73 +1188,71 @@ renameActionTests = testGroup "rename actions"
...
+        , "func x ="
+        , "  let y :: Int"
+        , "      y = x * 2"
+        , "  in y"
+        ]
+  , testUseTypeSignature "multi-line message"
+        [ "func :: _"
+        , "func x y = x + y"
+        ]
+        [ "func :: (Integer -> Integer -> Integer)"
```
Comment (michaelpj) [inline ghcide/test/exe/Main.hs:1215]: Hmmm, I guess it's not easy to tell if the hole is the entirety of the type signature? That seems like the remaining slightly bad case, and also probably a relatively common one...

Comment (michaelpj) [reply ghcide/test/exe/Main.hs:1215]: Anyway, that can be future work.

**Comment by AndreasPK** [comment]
Thanks for this!

**Pull request merged**